EIPS/eip-2225.md

```diff
@@ -0,0 +1,126 @@
+---
+eip: 2225
+title: Methods to Merge Ethereum Blockchains Together
+author: Wei Tang (@sorpaas)
+discussions-to: https://github.com/sorpaas/EIPs/issues/8
+status: Draft
+type: Standards Track
+category: Core
+created: 2019-08-02
+requires: 155, 1702
+---
+
+This EIP defines methods that can be adopted to merge two
+Ethereum-like blockchains together.
+
+Note that this EIP is purely an exploration of technical
+possibilities, and it does not mean the author endorse merging
+Ethereum and Ethereum Classic, or any other two blockchains together.
+
+## Simple Summary
+
+We define methods to merge multiple blockchains together, in a fully
+backward compatible way.
+
+## Abstract
+
+This defines how Ethereum merging can happen, with backward
+compatibility in mind. This is done by utilizing account versioning
+together with transaction replay protection. We also define three
+additional opcodes to allow merged chains to work across two sides of
+state.
+
+## Motivation
+
+There are many reasons why we may want to merge multiple Ethereum
+(eth1.x) blockchains together. For example, there may have been chain
+splits of a blockchain, where later the community issues are resolved,
+and we want to avoid a permanent split due to a temporary issue. It
+may also be two completely distinct blockchains that want to merge in
+order to share proof of work, or to improve inter-chain communication.
+
+## Specification
+
+The merge process happens at fork block, where we have a **merger**
+chain and a **mergee** chain. At fork height, a miner should mine both
+a merger chain's block, and a mergee chain's block. For those two
+blocks:
+
+* Both of them should have no transactions.
+* Merger chain's fork height block and mergee chain's fork height
+  block should have its state root processed as normal.
+* Merger chain's fork height block should set its extra data field to
+  mergee chain's fork height block's state root. The same for mergee
+  chain's fork height block.
```

Comment:
How can merger and mergee know the other's fork block height state root if they are waiting for the other to actually finish their blocks?

Reply:
Prior to the fork block, those miners that support the merge should maintain chain of both merger and mergee (as normal full nodes, meaning it should handle shadow reorg). On fork height, the miner broadcast two “cousin” blocks together, of both merger and mergee. Other nodes only accept those two cousin blocks if both of them are in the node’s local state. From this point on, every node who accepts the cousin fork blocks have the same view of the merger and mergee’s state root.

Reply:
This implies both chains have to reach their fork height at the same time so they know the state root from the other chain.  Given how hard it is to predict when a particular block height will be reached, that creates a significant potential for one chain to stall for quite a long time waiting for the other to catch up.

Reply:
Good point, but I suppose this can be fixed by only setting the fork block for merger chain, and let the miner decide mergee chain's head.

Reply:
Or, maybe, instead of using a single block as a "fork block" to use a 3 step process:

1. Merger chain reaches fork block 'A' and generates a fork block state root.
2. Mergee chain "sees" the merger chain's fork block 'A' state root and includes it in its fork block 'B' extra data field, and it generates its own state root.
3. Merger chain "sees" the mergee chain's fork block 'B' state root and includes it in its fork block 'C' extra data field.

So there would be three fork blocks A, B, and C (A and C in merger chain, and B in mergee chain). After 'C' the chains are merged.

Reply:
I think there may be fork choice issues in that. The reason why we have the setup in this EIP is that we ensure merger fork block and mergee fork block are produced at the same time.